muzzle/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/AgentCachingPoolStrategy.java

```diff
@@ -174,11 +174,13 @@ private static final class TypeCacheKey {
     private final int hashCode;
 
     TypeCacheKey(int loaderHash, WeakReference<ClassLoader> loaderRef, String className) {
-      this.loaderHash = loaderHash;
-      this.loaderRef = loaderRef;
+      // classes in java package are always loaded from boot loader
+      // set loader to boot loader to avoid creating multiple cache entries
+      this.loaderHash = className.startsWith("java.") ? BOOTSTRAP_HASH : loaderHash;
+      this.loaderRef = className.startsWith("java.") ? null : loaderRef;
```

Comment:
You should also add a nullcheck to `equals()` (and a `@Nullable` annotation on field) if you're making this nullable (NPE should not ever happen, but let's stay on the safer side).

Reply:
`null` was already a valid value for this field, it is used as a marker for boot loader. I think it could NPE if some class loader had the same hash code as the magic value used for boot loader hash.